### PR TITLE
Simplify/remove /persist/status/zedagent/*

### PR DESCRIPTION
**Pull request by eriknordmark**
# Description

We currently have a checkpoint of the protobuf config in /persist/checkpoint/lastconfig, which is signed by the controller. Its signature is verified before it is used, and then it is used to populate zedagent's publications even if there is no connection to the controller.
Thus the persistent publications (with files in /persist/status/zedagent) should not be needed, and getting rid of them simplifies analyzing any security impact unauthorized modifications will have to such files.

However, we need to have the ControllerCerts available since the CipherContext (used for object encryption) depend on those. This is adddressed by introducing a new /persist/checkpoint/controllercerts which contains the protobuf objects received from the controller. This file is then verified at boot (the same way as when we receive an update - that the certificate chain verifies all the way to the cert in /config/root-certificate.pem)

Both lastconfig and controllercerts have a .bak file which should ensure that even if there is a power outage when the file is written we will have a valid backup which we can use.

Note that more of the publications in /persist/status/zedagent need to be addressed.
Next is ConfigItemValueMap which might have some chicken and egg problems at bootup; need to have that published based on the checkpoint lastconfig as the agents start.

## How to test and validate this PR

Since we are touching code which relate to rolling the controller certificates that needs to be tested very carefully, including any corner case.
And since the purpose of the checkpoints are to allow the device (including datastore and WiFi credentials) and app instances (including with cloud-init) to boot even if there is no network, that needs to be carefully tested.

It is not clear whether we can test the corruption of /persist/checkpoint/lastconfig or /persist/checkpoint/controllercerts, but that is why we have the .bak files (to handle inopportune power outages while the checkpoint file(s) are updated.)

## Changelog notes

Removed extra copies of checkpointed state from /persist/status/<agentname> and instead solely rely on the signed protobuf encoded checkpoint in /persist/checkpoint/lastconfig. This required also checkpointing the controller certs (in /persist/checkpoint/controllercerts), and providing a backup copy of both checkpoints. The backups are needed to handle both the case of inopportune power outages when the checkpoints are written, and the fact that the lastconfig depends on the controllercerts yet those two separate checkpoint files are not updated atomically.

## PR Backports

Here is the list of current LTS branches (it should be always up to date):

- 16.0-stable To be backported.
- 14.5-stable To be backported.
- 13.4-stable No.

## Checklist

- [X] I've provided a proper description
- [X] I've added the proper documentation
- [X] I've tested my PR on amd64 device
- [ ] I've tested my PR on arm64 device
- [X] I've written the test verification instructions
- [X] I've set the proper labels to this PR

And the last but not least:

- [X] I've checked the boxes above, or I've provided a good reason why I didn't
  check them.

Please, check the boxes above after submitting the PR in interactive mode.

### CODE REVIEW

**Comment by europaul** [comment]
The last two commits confuse me a bit: is this still a work-in-progress?

**Review by europaul — COMMENTED** [review]
The changes look good to me in general - you caught a lot of details that I might have missed :) I also 100% support the dismantling of `upgradeconverter` package

**Comment by eriknordmark** [comment]
> The last two commits confuse me a bit: is this still a work-in-progress?

The last commit was incomplete so I removed that one (and removed/rewrote various XXX comments in the other commits). Thus there will still be some /persist/status/zedagent/ files related to EVE-k which can be handled separately.

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 1]
```diff
@@ -473,6 +473,15 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
...
 		zedagentCtx.agentMetrics)
 
+	// initialize ControllerCerts by reading from its /persist/checkpoint
+	// if that exists
+	initializeControllerCerts(zedagentCtx, ctrlClient)
+
+	// initialize a ConfigItemValueMap starting with the defaults,
+	// then parse only the ConfigItemValueMap parts from /persist/checkpoint/lastconfig,
+	// and then publish it.
+	initializeConfigItemValueMap(zedagentCtx, ctrlClient)
```
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/zedagent.go]: This initialization happens (by parseEdgeNodeClusterConfig calling publishEmptyENCC) when the checkpointed config is parsed at zedagent startup.

**Review by rene — APPROVED** [review]
Kick off tests....

File: pkg/pillar/cmd/client/client.go [hunk 2]
```diff
@@ -715,25 +712,85 @@ func fetchCertChain(ctrlClient *controllerconn.Client, tlsConfig *tls.Config, re
...
+}
+
+// Returns true if there is a change/difference by comparing the sorted
+// list of certs
+func compareControllerCertBytes(newCertChainBytes, prevCertChainBytes []byte) bool {
+	newKeys, _ := parseKeysFromControllerCerts(newCertChainBytes)
+	prevKeys, _ := parseKeysFromControllerCerts(prevCertChainBytes)
+	allNewKeys := strings.Join(newKeys, " ")
+	allPrevKeys := strings.Join(prevKeys, " ")
+	if allNewKeys != allPrevKeys {
```
Comment (milan-zededa) [inline pkg/pillar/cmd/client/client.go]: You can use this to avoid sort.Slice and strings.Join: https://github.com/lf-edge/eve/blob/master/pkg/pillar/utils/generics/generics.go#L31-L40

File: pkg/pillar/cmd/upgradeconverter/upgradeconverter.go [hunk 3]
```diff
@@ -28,22 +28,6 @@ const (
...
-	{
-		description: "Convert Global Settings to new format",
-		handlerFunc: convertGlobalConfig,
-	},
-	{
-		description: "Move ConfigItemValueMap from /persist/config to /persist/status",
-		handlerFunc: moveConfigItemValueMap,
-	},
-	{
-		description: "Move any configuration files from /config/GlobalConfig to /persist/status",
```
Comment (milan-zededa) [inline pkg/pillar/cmd/upgradeconverter/upgradeconverter.go:40]: Are we going to abandon the feature of injecting the initial global config into the installer? (aka the override.json for GlobalConfig)
It for example allows to enable ssh even before onboarding and use that for troubleshooting when console access or monitor is not available.
You can also enable pre-onboarding debug logs, which can be useful when troubleshooting some init code.

File: pkg/pillar/cmd/vaultmgr/vaultmgr.go [hunk 4]
```diff
@@ -224,7 +224,7 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
 		AgentName:     "zedagent",
 		MyAgentName:   agentName,
 		TopicImpl:     types.ConfigItemValueMap{},
-		Persistent:    true,
+		Persistent:    false,
```
Comment (milan-zededa) [inline pkg/pillar/cmd/vaultmgr/vaultmgr.go:227]: Many microservices wait for ConfigItemValueMap before they start doing any useful work.
Will this delay their startup?

Comment (milan-zededa) [reply pkg/pillar/cmd/vaultmgr/vaultmgr.go:227]: Never mind, I see that you handle this with `civmOnly`.

File: pkg/pillar/cmd/zedagent/handlecertconfig.go [hunk 5]
```diff
@@ -371,8 +411,15 @@ func requestCertsByURL(ctx *zedagentContext, certURL string, desc string,
 		return true
 	}
 
-	// write the signing cert to file
-	if err := ctrlClient.SaveServerSigningCert(signingCertBytes); err != nil {
+	// Update the checkpoint - note that since we've verified semantic
+	// difference in parsePublishControllerCerts this will not be triggered
+	// due to mere order differences in the protobuf contents
+	persist.MaybeSaveControllerCerts(log, rv.RespContents)
```
Comment (milan-zededa) [inline pkg/pillar/cmd/zedagent/handlecertconfig.go:417]: I see that both zedagent and client are saving controller certs. Could this create a race condition between the two? Would it be possible to simplify this and have one owner for this file?

Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handlecertconfig.go:417]: I'll double-check but zedclient runs before zedagent is active (zedagent waits for onboarded) and we haven't seen any issues with writing the current certificate files from both places.

File: pkg/pillar/cmd/upgradeconverter/upgradeconverter.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/upgradeconverter/upgradeconverter.go:40]: That was not the intent. But this needs to be done differently since there is no file in /persist/status/zedagent/ConfigItemValueMap we can write to.
I'll come up with something using a new file in /run which zedagent will use; if we only do this on first boot it shouldn't be hard.

@milan-zededa Have we suggested or documented any cases where a user would write into /config/GlobalConfig/ on an already booted system?

File: pkg/pillar/controllerconn/authen.go [hunk 6]
```diff
@@ -105,9 +106,29 @@ func (c *Client) removeAndVerifyAuthContainer(
...
+// The client captures the current chain of certs, which can be retrieved
+// using GetCertChainBytes() after the verification.
 func (c *Client) VerifyAuthContainerHeader(sm *zauth.AuthContainer) (
 	types.SenderStatus, error) {
-	err := c.LoadSavedServerSigningCert()
+	status, err := c.tryVerifyAuthContainerHeader(sm, false)
+	if status == types.SenderStatusCertMiss {
+		// Try backup
+		c.ClearServerCert()
+		status, err = c.tryVerifyAuthContainerHeader(sm, false)
```
[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/controllerconn/authen.go]: Fixed

File: pkg/pillar/cmd/client/client.go (hunk 2 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/client/client.go]: Nice. I'll change that.

File: pkg/pillar/cmd/upgradeconverter/upgradeconverter.go (hunk 3 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/upgradeconverter/upgradeconverter.go:40]: @milan-zededa I have finally re-implemented the handling of /config/GlobalConfig/ and /config/authorized_keys in zedagent (to replace the code in upgradeconverter) following the pattern for handing the bootstrap config.

However, there are two XXX questions in the last commit (which has those changes):
1. Putting the sha256 in /persist/ingested doesn't make much sense since we don't ingest these files into /persist - we should re-read them on boot until we get a config from the controller 
2. AFAICT it makes sense to not ingest those files, nor the bootstrap config, if we already have a /persist/checkpoint/lastconfig i.e., if we have been onboarded and received config from the controller.

Thoughts?

File: pkg/pillar/cmd/zedagent/zedagent.go [hunk 7]
```diff
@@ -416,28 +421,46 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
...
-	} else {
-		log.Warnf("GlobalConfig is missing, publishing default values")
-		zedagentCtx.globalConfig = *types.DefaultConfigItemValueMap()
-		err = zedagentCtx.pubGlobalConfig.Publish("global", zedagentCtx.globalConfig)
+	// Check if /persist/checkpoint/lastconfig exists.
+	// If not make sure we have defaults for ConfigItemValueMap.
+	// Does NOT publish yet since we want to get any bootstrap or
+	foundCheckpoint := initializeConfigItemValueMap(zedagentCtx)
+
+	// XXX should a bootstrap or /config/GlobalConfig file be able
```
Comment (milan-zededa) [inline pkg/pillar/cmd/zedagent/zedagent.go]: For bootstrap config we write its SHA to `/persist/ingested/bootstrap-config.sha` and use it to avoid loading it again on future device boots (this is why the function is called **maybe**LoadBootstrapConfig).
Could we use something similar for `/config/GlobalConfig` ?

Something like (execute on every boot):
1. if `/config/bootstrap-config.pb` exists -> run `maybeLoadBootstrapConfig`
2. if `/config/bootstrap-config.pb` does not exist, maybe ingest `/config/GlobalConfig` (if it has not been done yet, check SHA file presence)

But if we want to simplify and avoid too many persistent files, then using the checkpoint should work as well I think. In that case `maybeLoadBootstrapConfig` could be also simplified, dropping the part where we write `bootstrap-config.sha`

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 8]
```diff
@@ -328,13 +338,145 @@ func maybeLoadBootstrapConfig(getconfigCtx *getconfigContext) {
...
 	getconfigCtx.ledBlinkCount = types.LedBlinkInvalidBootstrapConfig
 }
 
+// Load /config/GlobalConfig/ json file with ConfigItemValueMap provided that:
+//   - has not been loaded before (incl. previous device boots)
+//
+// The function will only load the ConfigItemValueMap items into
+// zedagentCtx,globalConfig. The caller is responsible for publishing those.
+func maybeLoadGlobalConfig(getconfigCtx *getconfigContext) {
+	//  Check if bootstrap config has been already loaded.
```
Comment (milan-zededa) [inline pkg/pillar/cmd/zedagent/handleconfig.go]: Check if bootstrap -> Check if global config

File: pkg/pillar/cmd/zedagent/handleconfig.go [hunk 9]
```diff
@@ -328,13 +338,145 @@ func maybeLoadBootstrapConfig(getconfigCtx *getconfigContext) {
...
+		// and as a result load it again with the next boot.
+		// However, with config timestamp preventing accidental revert
+		// to an older configuration, this should not be a problem.
+		configSha = nil
+	}
+	// Mark global config as processed by storing SHA hash of its content
+	// under /persist/ingested/ directory.
+	// We do this even even if the unmarshalling (or anything else) below fails.
+	// This is to avoid repeated failing attempts to load invalid config on each boot.
+	// XXX but not ingested to disk!!!
```
Comment (milan-zededa) [inline pkg/pillar/cmd/zedagent/handleconfig.go]: Ah, now I understand what you mean -- **GlobalConfig is no longer a persistent publication**, so we may need to re-ingest it multiple times until configuration is received from the controller.

In that case, I suggest we stop writing the SHA file here and instead continue ingesting `/config/GlobalConfig/` as long as there is neither bootstrap configuration nor a controller config checkpoint present.

We will run into the same issue with global config properties defined in the bootstrap configuration -- those should also be re-ingested on every boot until we receive configuration from the controller. Given that, it might make sense to stop using the `bootstrap-config.sha` file as well.

I don’t expect any issues from repeatedly re-ingesting the bootstrap configuration on each boot until the first controller-provided configuration is available.

File: pkg/pillar/cmd/zedagent/zedagent.go (hunk 7 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/zedagent.go]: Disregard this comment, see my other comments below.

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 9 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: @milan-zededa I put some explicit checks in nim.go to make sure we don't re-ingest after we have a /persist/checkpoint/lastconfig. Please take a look at those checks.

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 8 above)
Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: Fixed.

**Review by milan-zededa — APPROVED** [review]

File: pkg/pillar/cmd/zedagent/handleconfig.go (hunk 9 above)
Comment (milan-zededa) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: Done, LGTM

File: pkg/pillar/cmd/zedagent/handleconfig.go
Comment (milan-zededa) [inline pkg/pillar/cmd/zedagent/handleconfig.go]: Should we strip the "maybe" prefix from the function name now that we reload bootstrap config on every boot?

Comment (eriknordmark) [reply pkg/pillar/cmd/zedagent/handleconfig.go]: Done
